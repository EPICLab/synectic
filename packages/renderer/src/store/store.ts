import {combineReducers, configureStore} from '@reduxjs/toolkit';
import branchesReducer from './slices/branches';
import cardsReducer from './slices/cards';
import commitsReducer from './slices/commits';
import filetypesReducer from './slices/filetypes';
import metafilesReducer from './slices/metafiles';
import modalsReducer from './slices/modals';
import reposReducer from './slices/repos';
import stacksReducer from './slices/stacks';
import {listenerMiddleware} from './listenerMiddleware';

export const rootReducer = combineReducers({
  branches: branchesReducer,
  cards: cardsReducer,
  commits: commitsReducer,
  filetypes: filetypesReducer,
  metafiles: metafilesReducer,
  modals: modalsReducer,
  repos: reposReducer,
  stacks: stacksReducer,
});

const store = configureStore({
  reducer: rootReducer,
  devTools: true,
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware().concat([listenerMiddleware.middleware]),
});

export type RootState = ReturnType<typeof rootReducer>;
export type AppDispatch = typeof store.dispatch;

export default {store: store};
