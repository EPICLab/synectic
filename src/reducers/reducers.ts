import { combineReducers } from 'redux';
import {
  // ADD_TODO,
  // TOGGLE_TODO,
  SET_VISIBILITY_FILTER,
  VisibilityFilters,
  ActionType
} from '../actions/actions';

const { SHOW_ALL } = VisibilityFilters;

// type TodoItem = {
//   text: string;
//   completed: boolean;
// }

// type TodoState = {
//   visibilityFilter: string;
//   todos: TodoItem[];
// }

function visibilityFilter(state = SHOW_ALL, action: ActionType) {
  switch (action.type) {
    case SET_VISIBILITY_FILTER:
      return action.filter;
    default:
      return state;
  }
}

// function todos(state: TodoItem[] = [], action: ActionType) {
//   switch (action.type) {
//     case ADD_TODO:
//       return [
//         ...state,
//         {
//           text: action.text,
//           completed: false
//         }
//       ];
//     case TOGGLE_TODO:
//       return state.map((todo, index) => {
//         if (index === action.index) {
//           return Object.assign({}, todo, {
//             completed: !todo.completed
//           });
//         }
//         return todo;
//       });
//     default:
//       return state;
//   }
// }

const todoApp = combineReducers({
  visibilityFilter
});

export default todoApp;