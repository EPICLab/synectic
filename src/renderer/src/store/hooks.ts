import { createAsyncThunk } from '@reduxjs/toolkit';
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import type { RootState, AppDispatch } from './store';

export const createAppAsyncThunk = createAsyncThunk.withTypes<{
  state: RootState;
  dispatch: AppDispatch;
  rejectValue: string;
}>();

export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
