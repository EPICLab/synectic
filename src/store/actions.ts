import { UUID, Repository, Card, Stack, Metafile, Filetype } from '../types';

export enum ActionKeys {
  INITIALIZE_CANVAS,
  ADD_CARD, REMOVE_CARD, UPDATE_CARD,
  ADD_STACK, REMOVE_STACK, UPDATE_STACK,
  ADD_FILETYPE, REMOVE_FILETYPE, UPDATE_FILETYPE,
  ADD_METAFILE, REMOVE_METAFILE, UPDATE_METAFILE,
  ADD_REPO, REMOVE_REPO, UPDATE_REPO
}

export type Actions =
  | InitializeCanvasAction
  | AddCardAction
  | RemoveCardAction
  | UpdateCardAction
  | AddStackAction
  | RemoveStackAction
  | UpdateStackAction
  | AddFiletypeAction
  | RemoveFiletypeAction
  | UpdateFiletypeAction
  | AddMetafileAction
  | RemoveMetafileAction
  | UpdateMetafileAction
  | AddRepoAction
  | RemoveRepoAction
  | UpdateRepoAction;

type InitializeCanvasAction = {
  type: ActionKeys.INITIALIZE_CANVAS;
};

type AddCardAction = {
  type: ActionKeys.ADD_CARD;
  id: UUID;
  card: Card;
};

type RemoveCardAction = {
  type: ActionKeys.REMOVE_CARD;
  id: UUID;
};

type UpdateCardAction = {
  type: ActionKeys.UPDATE_CARD;
  id: UUID;
  card: Partial<Card>;
};

type AddStackAction = {
  type: ActionKeys.ADD_STACK;
  id: UUID;
  stack: Stack;
}

type RemoveStackAction = {
  type: ActionKeys.REMOVE_STACK;
  id: UUID;
};

type UpdateStackAction = {
  type: ActionKeys.UPDATE_STACK;
  id: UUID;
  stack: Partial<Stack>;
};

type AddFiletypeAction = {
  type: ActionKeys.ADD_FILETYPE;
  id: UUID;
  filetype: Filetype;
};

type RemoveFiletypeAction = {
  type: ActionKeys.REMOVE_FILETYPE;
  id: UUID;
};

type UpdateFiletypeAction = {
  type: ActionKeys.UPDATE_FILETYPE;
  id: UUID;
  filetype: Partial<Filetype>;
};

type AddMetafileAction = {
  type: ActionKeys.ADD_METAFILE;
  id: UUID;
  metafile: Metafile;
};

type RemoveMetafileAction = {
  type: ActionKeys.REMOVE_METAFILE;
  id: UUID;
};

type UpdateMetafileAction = {
  type: ActionKeys.UPDATE_METAFILE;
  id: UUID;
  metafile: Partial<Metafile>;
};

type AddRepoAction = {
  type: ActionKeys.ADD_REPO;
  id: UUID;
  repo: Repository;
};

type RemoveRepoAction = {
  type: ActionKeys.REMOVE_REPO;
  id: UUID;
}

type UpdateRepoAction = {
  type: ActionKeys.UPDATE_REPO;
  id: UUID;
  repo: Partial<Repository>;
};