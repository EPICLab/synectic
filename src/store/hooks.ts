import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import type { RootState, AppDispatch } from './store';

export type AppThunkAPI<R = unknown> = { state: RootState, dispatch: AppDispatch, rejectValue: R };

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;