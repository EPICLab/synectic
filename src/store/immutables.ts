/**
 * Immutably copy the values of all enumerable own properties from old object 
 * and map new values onto it in a new target object.
 * @param oldObject The initial source object from which to copy properties.
 * @param newValues The new value object from which to copy properties.
 * @return New object containing all enumerable own properties from both params.
 */
export const updateObject = <T>(oldObject: T, newValues: Partial<T>): T => {
  return Object.assign({}, oldObject, newValues);
};

/**
 * Immutably append a new item to an array by constructing a new combined array.
 * @param oldArray The initial source Array object.
 * @param newItem A new item to be appended to array.
 * @returns New array containing all items from old array and including new item.
 */
export const addItemInArray = <T>(oldArray: T[], newItem: T): T[] => [...oldArray, newItem];

/**
 * Immutably remove an element from an array by producing a new reduced array.
 * @param array The initial source Array object.
 * @param item A target item contained within the array.
 * @returns New array containing all elements from array excluding target item.
 */
export const removeItemInArray = (array: string[], item: string): string[] => {
  return array.filter(element => element !== item);
}

/**
 * Immutably append a new item to a map by constructing a new combined map.
 * @param map The initial source key-value map object.
 * @param newItem A new item to be appended to map.
 * @returns New map containing all items from old map and including new item.
 */
export const addItemInMap = <T extends { id: string }>(map: { [id: string]: T }, newItem: T): { [id: string]: T } => {
  const updatedItems: { [id: string]: T } = {};
  for (const k in map) {
    updatedItems[k] = map[k];
  }
  updatedItems[newItem.id] = newItem;
  return updatedItems;
};

/**
 * Immutably remove an item from a map by producing a new reduced map.
 * @param map The initial source key-value map object.
 * @param itemId An id associated with an item contained in the map.
 * @returns New map containing all items from map excluding item with matching id.
 */
export const removeItemInMap = <T extends { id: string }>(map: { [id: string]: T }, itemId: string): { [id: string]: T } => {
  return Object.keys(map).reduce((items: { [id: string]: T }, id) => {
    if (id !== itemId) {
      items[id] = map[id];
    }
    return items;
  }, {});
}

/**
 * Immutably remove items from map using specified filter function.
 * @param map The initial source key-value map object.
 * @param filterFn Filter function that returns true for each item in map that 
 * meets conditions specified in function.
 */
export const removeMatchesInMap = <T>(map: { [id: string]: T }, filterFn: (item: T) => boolean): { [id: string]: T } => {
  return Object.keys(map).reduce((items: { [id: string]: T }, id) => {
    if (!filterFn(map[id])) {
      items[id] = map[id];
    }
    return items;
  }, {});
};

/**
 * Immutably filters items from map using specified filter function and applies a 
 * callback function to update each matching item.
 * @param map The initial source key-value map object.
 * @param filterFn Filter function that returns true for each item in map that 
 * meets conditions specified in function.
 * @param updateItemCallback Callback function to apply towards items that meet 
 * filter function predicates.
 */
export const updateMatchesInMap = <T>(map: { [id: string]: T }, filterFn: (item: T) => boolean, updateItemCallback: (item: T) => T): { [id: string]: T } => {
  const updatedItems: { [id: string]: T } = {};
  for (const k in map) {
    if (filterFn(map[k])) {
      updatedItems[k] = updateItemCallback(map[k]);
    } else {
      updatedItems[k] = map[k];
    }
  }
  return updatedItems;
};

/**
 * Immutably update a specific item in a key-value map based on item id 
 * and applying a callback function to that item.
 * @param map The initial source key-value map object.
 * @param itemId An id associated with an item contained in the map.
 * @param updateItemCallback Callback function to apply towards item with matching id.
 * @returns New map containing all items from map including updated item.
 */
export const updateItemInMapById = <T extends { id: string }>(map: { [id: string]: T }, itemId: string, updateItemCallback: (item: T) => T): { [id: string]: T } => {
  return updateMatchesInMap(map, (item => item.id === itemId), updateItemCallback);
};

