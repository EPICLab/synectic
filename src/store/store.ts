import { combineReducers, configureStore } from '@reduxjs/toolkit';
import stacksReducer from './slices/stacks';
import cardsReducer from './slices/cards';
import filetypesReducer from './slices/filetypes';
import metafilesReducer from './slices/metafiles';
import reposReducer from './slices/repos';
import modalsReducer from './slices/modals';
import { FLUSH, PAUSE, PERSIST, persistReducer, persistStore, PURGE, REGISTER, REHYDRATE } from 'redux-persist';
import storage from 'redux-persist/lib/storage';

export const rootReducer = combineReducers({
    stacks: stacksReducer,
    cards: cardsReducer,
    filetypes: filetypesReducer,
    metafiles: metafilesReducer,
    repos: reposReducer,
    modals: modalsReducer
});

const persistConfig = {
    key: 'root',
    storage: storage,
    version: 0,
    blacklist: ['filetypes']
}

export const persistedReducer = persistReducer(persistConfig, rootReducer);

const store = configureStore({
    reducer: persistedReducer,
    middleware: (getDefaultMiddleware) => [
        ...getDefaultMiddleware({
            serializableCheck: {
                ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER]
            }
        }),
    ]
});
const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default ({ store: store, persistor: persistor });