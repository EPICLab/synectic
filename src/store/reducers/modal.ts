import type { Modal } from '../../types';
import { Action, ActionKeys } from '../actions';
import { addItemInMap, removeItemInMap } from '../immutables';

export const modalReducer = (state: { [id: string]: Modal } = {}, action: Action): { [id: string]: Modal } => {
  switch (action.type) {
    case ActionKeys.ADD_MODAL: {
      return addItemInMap(state, action.modal);
    }
    case ActionKeys.REMOVE_MODAL: {
      return removeItemInMap(state, action.id);
    }
    default:
      return state;
  }
}