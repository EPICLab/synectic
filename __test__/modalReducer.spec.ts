import type { Modal } from '../src/types';
import { modalReducer } from '../src/store/reducers/modal';
import { ActionKeys } from '../src/store/actions';

describe('modalReducer', () => {
  const modals: { [id: string]: Modal } = {
    '92': {
      id: '92',
      type: 'Error',
      subtype: 'RepositoryMissingError',
      target: '23',
      options: { message: 'Repository missing for metafile \'sampleUser/myRepo\'' }
    }
  }

  const newModal: Modal = {
    id: '94',
    type: 'Error',
    subtype: 'HandlerMissingError',
    target: '19',
    options: { message: 'Metafile \'sampleUser/myRepo\' missing handler to resolve filetype: \'TypeScript\'' }
  }

  it('modalReducer returns default state when current state is blank', () => {
    const newModals = modalReducer(undefined, { type: ActionKeys.ADD_MODAL, id: newModal.id, modal: newModal });
    expect(Object.keys(newModals)).toHaveLength(1);
    expect(newModals).toMatchSnapshot();
  });

  it('modalReducer appends a new modal to state on action ADD_ERROR', () => {
    const addedModals = modalReducer(modals, { type: ActionKeys.ADD_MODAL, id: newModal.id, modal: newModal });
    expect(Object.keys(addedModals)).toHaveLength(2);
    expect(addedModals).toMatchSnapshot();
  });

  it('modalReducer removes an modal from state on action REMOVE_ERROR', () => {
    const matchedModals = modalReducer(modals, { type: ActionKeys.REMOVE_MODAL, id: '92' });
    expect(Object.keys(matchedModals)).toHaveLength(0);
  });

  it('modalReducer resolves non-matching modal in state on action REMOVE_ERROR', () => {
    const nonMatchedModals = modalReducer(modals, { type: ActionKeys.REMOVE_MODAL, id: newModal.id });
    expect(Object.keys(nonMatchedModals)).toHaveLength(Object.keys(modals).length);
  });

});