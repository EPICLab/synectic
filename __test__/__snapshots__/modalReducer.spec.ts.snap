// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modalReducer modalReducer appends a new modal to state on action ADD_ERROR 1`] = `
Object {
  "92": Object {
    "id": "92",
    "options": Object {
      "message": "Repository missing for metafile 'sampleUser/myRepo'",
    },
    "subtype": "RepositoryMissingError",
    "target": "23",
    "type": "Error",
  },
  "94": Object {
    "id": "94",
    "options": Object {
      "message": "Metafile 'sampleUser/myRepo' missing handler to resolve filetype: 'TypeScript'",
    },
    "subtype": "HandlerMissingError",
    "target": "19",
    "type": "Error",
  },
}
`;

exports[`modalReducer modalReducer returns default state when current state is blank 1`] = `
Object {
  "94": Object {
    "id": "94",
    "options": Object {
      "message": "Metafile 'sampleUser/myRepo' missing handler to resolve filetype: 'TypeScript'",
    },
    "subtype": "HandlerMissingError",
    "target": "19",
    "type": "Error",
  },
}
`;
